deploy/helm/kubecf/templates/_config.tpl

```diff
@@ -0,0 +1,290 @@
+| (without the quotes, of course).
+|
+| - A nil (missing) condition is always true.
+|
+| - A boolean condition is the value of the $condition itself.
+|
+| - Otherwise $condition must be a string. All spaces will be removed first.
+|
+| - The string can be a list of "||" separated OR terms. The condition is true
+|   if any of the terms are true.
+|
+| - An OR term can be a list of "&&" separated AND terms. The OR term is true
+|   if all of the AND terms are true.
+|
+| - An AND term is either the string "true" or "false", or it will be evaluated
+|   by looking up it's value in $.kubecf.config.
+|
+| - An AND term may be prefixed by "!" in which case its value is negated.
+|
+| - An AND term may also be a condition expression enclosed in parenthesis.
+|
+| - None of the terms can be an empty string.
+|
+| Example: "!features.foo.enabled && (features.bar.enabled || features.baz.enabled)"
+|
+| Usage examples: This function is used to evaluate the keys of the
+| "deprecation" hash earlier in this file, and the "release.condition" values
+| in assets/operations/releases.yaml.
+==========================================================================================
+*/}}
+{{- define "_config.condition" }}
+  {{- $root := index . 0 }}
+  {{- $condition := index . 1 }}
+
+  {{- if kindIs "invalid" $condition }}
+    {{- /* The absence of a condition (nil) is unconditionally true */}}
+    {{- $condition = true }}
+  {{- end }}
+
+  {{- if kindIs "bool" $condition }}
+    {{- ternary "true" "false" $condition }}
+
+  {{- else }}
+    {{- /* Count the number of left parenthesis to determine the number of groups */}}
+    {{- range $_ := until (sub (len (splitList "(" $condition)) 1 | int) }}
+      {{- /* Find left inner-most group, evaluate it, and replace the expression with the value */}}
+      {{- $group := regexFind "\\([^\\)]+\\)" $condition }}
+      {{- /* There may be fewer groups than left parens if some groups turn out to be identical */}}
+      {{- /* E.g. "((to.be) || !(to.be))" will collapse to "(true || !true)" after the 1st iteration */}}
+      {{- if $group }}
+        {{- $value := include "_config.condition" (list $root ($group | trimPrefix "(" | trimSuffix ")")) }}
+        {{- $condition = replace $group $value $condition }}
+      {{- end }}
+    {{- end }}
+
+    {{- /* Evaluate the remaining expression based on operator precedence: NOT (highest), AND, OR (lowest) */}}
+    {{- $or_value := false }}
+    {{- range $or_term := splitList "||" (nospace $condition) }}
+      {{- $and_value := true }}
+      {{- range $and_term := splitList "&&" $or_term }}
+        {{- $term := trimPrefix "!" $and_term }}
+        {{- /* The term is either literally "true" or "false", or must be looked up */}}
+        {{- $value := true }}
+        {{- if ne $term "true" }}
+          {{- if eq $term "false" }}
+            {{- $value = false }}
+          {{- else }}
+            {{- $value = include "_config.lookup" (list $root $term) }}
```

Comment:
No, I don't expect a bool; I expect a string that represents the zero-ness of the object at that location.

But you make a good point, this function would return the string `0` for an integer zero, and while the integer would be false, the string becomes true, so this is inconsistent with how go templating handles logical expressions otherwise.

I need to think about it some more and will refactor.

Reply:
So it turns out that `_config.condition` works correctly for a property with an integer `0` value, but only because `_config.lookup` returns the empty string for properties that contain the zero value for their type. I consider that a bug, so I guess I need some tests for `_config.lookup` too...

Reply:
Fixed in 1dc62f0f, including tests.